src/main/java/spoon/support/visitor/SubInheritanceHierarchyResolver.java

```diff
@@ -0,0 +1,230 @@
+	 *
+	 * 	If this method is called again with same input and configuration, nothing in sent to outputConsumer
+	 * @param outputConsumer the consumer for found sub types
+	 */
+	public <T extends CtTypeInformation> void forEachSubTypeInPackage(final CtConsumer<T> outputConsumer) {
+		/*
+		 * Set of qualified names of all visited types, independent on whether they are sub types or not.
+		 */
+		final Set<String> allVisitedTypeNames = new HashSet<>();
+		/*
+		 * the queue of types whose super inheritance hierarchy we are just visiting.
+		 * They are potential sub types of an `targetSuperTypes`
+		 */
+		final Deque<CtTypeReference<?>> currentSubTypes = new ArrayDeque<>();
+		//algorithm
+		//1) query step: scan input package for sub classes and sub interfaces
+		final CtQuery q = inputPackage.map(new CtScannerFunction());
+		//2) query step: visit only required CtTypes
+		if (includingInterfaces) {
+			//the client is interested in sub inheritance hierarchy of interfaces too. Check interfaces, classes, enums, Annotations, but not CtTypeParameters.
+			q.select(typeFilter);
+		} else {
+			//the client is not interested in sub inheritance hierarchy of interfaces. Check only classes and enums.
+			q.select(classFilter);
+		}
+		/*
+		 * 3) query step: for each found CtType, visit it's super inheritance hierarchy and search there for a type which is equal to one of targetSuperTypes.
+		 * If found then all sub types in hierarchy (variable `currentSubTypes`) are sub types of targetSuperTypes. So return them
+		 */
+		q.map(new SuperInheritanceHierarchyFunction()
+			//if there is any interface between `targetSuperTypes`, then we have to check superInterfaces too
+			.includingInterfaces(hasSuperInterface)
+			//internally it works always with references
+			.returnTypeReferences(true)
+			.failOnClassNotFound(failOnClassNotFound)
+			/*
+			 * listen for types in super inheritance hierarchy
+			 * 1) to collect `currentSubTypes`
+			 * 2) to check if we have already found a targetSuperType
+			 * 3) if found then send `currentSubTypes` to `outputConsumer` and skip visiting of further super types
+			 */
+			.setListener(new CtScannerListener() {
+				@Override
+				public ScanningMode enter(CtElement element) {
+					CtTypeReference<?> typeRef = (CtTypeReference<?>) element;
+					String qName = typeRef.getQualifiedName();
+					if (targetSuperTypes.contains(qName)) {
+						/*
+						 * FOUND! we are in super inheritance hierarchy, which extends from an searched super type(s).
+						 * All `currentSubTypes` are sub types of searched super type
+						 */
+						while (currentSubTypes.size() > 0) {
+							typeRef = currentSubTypes.pop();
+							/*
+							 * Send them to outputConsumer and add then as targetSuperTypes too, to perform faster with detection of next sub types.
+							 */
+							if (!targetSuperTypes.contains(typeRef.getQualifiedName())) {
+								targetSuperTypes.add(typeRef.getQualifiedName());

```

Comment:
```java
if (!targetSuperTypes.contains(typeRef.getQualifiedName())) {
   targetSuperTypes.add(typeRef.getQualifiedName());
```
can be replaced by
```java
if (targetSuperTypes.add(typeRef.getQualifiedName())) {
```
You have seen that in my origin code, and you changed it, so you probably think it as not understandable. So I at least suggest
```java
String qName = typeRef.getQualifiedName();
if (!targetSuperTypes.contains(qName)) {
   targetSuperTypes.add(qName);
```
... spoon is already not fastest so why to wast time again?
I have feeling (not measured) like this code is in middle of often repeated algorithm. I plan to refactor 10 000 methods of 6000 classes and I need to search many many inheritance hiearchies of all these classes for all methods ... it can end with a big number of calls ;-)